volume-providers/vpc/provider/get_volume_attachment.go

```diff
@@ -11,49 +11,84 @@
+	} else {
+		// Get volume attachment by Volume ID. This is inefficient operation which requires iteration over volume attachment list
+		volumeAttachmentResponse, err = vpcs.getVolumeAttachmentByVolumeID(volumeAttachment)
+	}
+	vpcs.Logger.Info("Volume attachment response", zap.Reflect("volumeAttachmentResponse", volumeAttachmentResponse), zap.Error(err))
+	return volumeAttachmentResponse, err
+
+}
+
+func (vpcs *VPCSession) getVolumeAttachmentByID(volumeAttachmentRequest models.VolumeAttachment) (*provider.VolumeAttachmentResponse, error) {
+	vpcs.Logger.Debug("Entry of getVolumeAttachmentByID()")
+	defer vpcs.Logger.Debug("Exit from getVolumeAttachmentByID()")
+	vpcs.Logger.Info("Getting VolumeAttachment from VPC provider...")
+	var err error
+	var volumeAttachmentResult *models.VolumeAttachment
+	err = retry(vpcs.Logger, func() error {
+		volumeAttachmentResult, err = vpcs.Apiclient.VolumeAttachService().GetVolumeAttachment(&volumeAttachmentRequest, vpcs.Logger)
+		return err
+	})
+	if err != nil {
+		// API call is failed
+		userErr := userError.GetUserError(string(userError.VolumeAttachFindFailed), err, volumeAttachmentRequest.Volume.ID, *volumeAttachmentRequest.InstanceID)
+		return nil, userErr
+	}
+	volumeAttachmentResponse := volumeAttachmentResult.ToVolumeAttachmentResponse()
+	vpcs.Logger.Info("Successfuly retrived volume attachment", zap.Reflect("volumeAttachmentResponse", volumeAttachmentResponse))
+	return volumeAttachmentResponse, err
+}
+
+func (vpcs *VPCSession) getVolumeAttachmentByVolumeID(volumeAttachmentRequest models.VolumeAttachment) (*provider.VolumeAttachmentResponse, error) {
+	vpcs.Logger.Debug("Entry of getVolumeAttachmentByVolumeID()")
+	defer vpcs.Logger.Debug("Exit from getVolumeAttachmentByVolumeID()")
 	vpcs.Logger.Info("Getting VolumeAttachmentList from VPC provider...")
 	var volumeAttachmentList *models.VolumeAttachmentList
+	var err error
 	err = retry(vpcs.Logger, func() error {
-		volumeAttachmentList, err = vpcs.Apiclient.VolumeAttachService().ListVolumeAttachment(&volumeAttachment, vpcs.Logger)
+		volumeAttachmentList, err = vpcs.Apiclient.VolumeAttachService().ListVolumeAttachment(&volumeAttachmentRequest, vpcs.Logger)
 		return err
 	})
 	if err != nil {
 		// API call is failed
-		userErr := userError.GetUserError(string(userError.VolumeAttachFindFailed), err, volumeAttachmentRequest.VolumeID, volumeAttachmentRequest.InstanceID)
+		userErr := userError.GetUserError(string(userError.VolumeAttachFindFailed), err, volumeAttachmentRequest.Volume.ID, *volumeAttachmentRequest.InstanceID)
 		return nil, userErr
 	}
 	// Iterate over the volume attachment list for given instance
 	for _, volumeAttachmentItem := range volumeAttachmentList.VolumeAttachments {
 		// Check if volume ID is matching with requested volume ID
-		if volumeAttachmentItem.Volume.ID == volumeAttachmentRequest.VolumeID {
+		if volumeAttachmentItem.Volume.ID == volumeAttachmentRequest.Volume.ID {
```

Comment:
VolumeAttachService().ListVolumeAttachment method itself should provide one record ? why we want to match with `volumeID` as one volume can be attached to one instance only.

Reply:
Multiple volumes can be attached to one instance